ord_schema/templating.py

```diff
@@ -63,20 +63,65 @@ def _escape(string):
     return string.replace('\\', '\\\\')
 
 
+def _is_null(value):
+    """Returns whether a value is null."""
+    return pd.isnull(value) or value == 'nan'
+
+
 def _replace(string, substitutions):
     """Performs substring substitutions according to a dictionary.
 
+    If any pattern has a null replacement value (i.e. this is an empty cell in
+    the spreadsheet), apply the following edits in the given order:
+
+        1. If after templating, any identifier does not have a defined value,
+           remove the identifier.
+        2. If a compound doesn't have any identifiers, remove that compound.
+        3. If a compound amount is NaN, remove that compound.
+        4. If an input has no components, remove that input.
+
+    These edits are logged for easier error checking.
+
+    Note that these edits are focused on *inputs* and are not intended to be
+    exhaustive (e.g. by covering all Compound messages anywhere in Reaction).
+
     Args:
         string: A string whose contents should be modified.
         substitutions: A dictionary where each (key, value) pair defines
             a substring to replace and what its replacement should be.
 
     Returns:
-        The modified string.
+        Reaction message with substitutions filled in.
+
+    Raises:
+        ValueError: If the substituted reaction template cannot be parsed.
     """
-    pattern = re.compile('|'.join(map(re.escape, substitutions.keys())))
-    return pattern.sub(lambda match: _escape(substitutions[match.group(0)]),
-                       string)
+    check_null = False
+    for pattern, value in substitutions.items():
+        if pd.isnull(value):
+            check_null = True
+        string = string.replace(pattern, _escape(str(value)), 1)
```

Comment:
Why only replace a single instance? I could imagine a situation in which multiple values need to be replaced. For example,
- During a solvent screen, multiple inputs are prepared as stock solutions in the same `$solvent_smiles$` that varies between experiments.
- A `$compound_name$` appears in an identifier and also in an unstructured field, like the procedure details

Reply:
That's fair; updated.